k9mail-library/src/main/java/com/fsck/k9/mail/transport/smtp/SmtpTransport.java

```diff
@@ -755,6 +755,68 @@ private String readCommandResponseLine(List<String> results) throws IOException
         return line;
     }
 
+    private void executePipelinedCommands(Queue<String> pipelinedCommands) throws IOException {
+        for (String command : pipelinedCommands) {
+            writeLine(command, false);
+        }
+    }
+
+    private void readPipelinedResponse(Queue<String> pipelinedCommands) throws IOException, MessagingException {
+        int noOfPipelinedResponse = pipelinedCommands.size();
+        String responseLine = null;
+        List<String> results = new ArrayList<>();
+        while (noOfPipelinedResponse > 0) {
+            results.clear();
+            responseLine = readCommandResponseLine(results);
+            try {
+                responseLineToCommandResponse(responseLine, results);
+
+            } catch (NegativeSmtpReplyException exception) {
+                //continue reading response till DATA response .
+                Log.d(LOG_TAG, "SMTP <<< " + exception.getReplyCode() + exception.getReplyText());
+
+            } catch (MessagingException exception) {
+                //continue reading response till DATA response .
+
+            }
+            noOfPipelinedResponse-- ;
+        }
+
+        try {
+            responseLineToCommandResponse(responseLine, results);
```

Comment:
I think the way it currently works has a subtle bug: the `results` list isn't cleared for the second time the command is run. Please use the "if (last in loop)" pattern above, I don't think it's less readable, but it avoids this type of inconsistency errors. :)

Otherwise I'm fine with this PR, nice work! :+1: